bionic/provider.py

```diff
@@ -90,27 +90,38 @@ def get_source_func(self):
     def copy(self):
         raise NotImplementedError()
 
+    @property
+    def entity_names(self):
+        return [
+            name for dnode in self.attrs.dnodes for name in dnode.all_entity_names()
+        ]
+
     def __repr__(self):
-        return f"{self.__class__.__name__}{tuple(self.attrs.names)!r}"
+        descriptors = tuple(dnode.to_descriptor() for dnode in self.attrs.dnodes)
+        return f"{self.__class__.__name__}{descriptors!r}"
 
 
 class ValueProvider(BaseProvider):
     @classmethod
     def from_single_value_providers(cls, value_providers):
+        names = []
         for provider in value_providers:
             assert isinstance(provider, ValueProvider)
-            assert len(provider.attrs.names) == 1
+            assert len(provider.attrs.dnodes) == 1
             assert not provider._has_any_values
 
-        names = [provider.attrs.names[0] for provider in value_providers]
+            (dnode,) = provider.attrs.dnodes
+            names.append(dnode.to_entity_name())
 
         return ValueProvider(names)
 
     def __init__(self, names):
+        dnodes = [entity_dnode_from_descriptor(name) for name in names]
         super(ValueProvider, self).__init__(
-            attrs=ProviderAttributes(names=names, changes_per_run=False,),
+            attrs=ProviderAttributes(dnodes=dnodes, changes_per_run=False,),
         )
 
+        self.names = names
```

Comment:
Wdyt about using ``self.attrs.entity_names` instead of storing the names here?

Reply:
I don't feel strongly either way -- happy to change it.